abstract.go

```diff
@@ -187,6 +195,105 @@ func scrapeDiscoveryJob(job job, tagsOnMetrics exportedTagsOnMetrics, clientTag
 	return awsInfoData, cw
 }
 
+func scrapeDiscoveryJobUsingMetricData(job job, tagsOnMetrics exportedTagsOnMetrics, clientTag tagsInterface, clientCloudwatch cloudwatchInterface) (awsInfoData []*tagsData, cw []*cloudwatchData) {
+	mux := &sync.Mutex{}
+	var wg sync.WaitGroup
+	var getMetricDatas []cloudwatchData
+	var length int
+
+	if job.Length == 0 {
+		length = 120
+	} else {
+		length = job.Length
+	}
+
+	resources, err := clientTag.get(job)
+	if err != nil {
+		log.Println("Couldn't describe resources: ", err.Error())
+		return
+	}
+	commonJobDimensions := getAwsDimensions(job)
+
+	for i := range resources {
+		resource := resources[i]
+		mux.Lock()
+		awsInfoData = append(awsInfoData, resource)
+		mux.Unlock()
+		metricTags := resource.metricTags(tagsOnMetrics)
+		dimensions := detectDimensionsByService(resource.Service, resource.ID, clientCloudwatch)
+		for _, commonJobDimension := range commonJobDimensions {
+			dimensions = append(dimensions, commonJobDimension)
+		}
+
+		wg.Add(len(job.Metrics))
+		go func() {
+			for j := range job.Metrics {
+				metric := job.Metrics[j]
+				dimensions = addAdditionalDimensions(dimensions, metric.AdditionalDimensions)
+				resp := getMetricsList(dimensions, resource.Service, metric, clientCloudwatch)
+				defer wg.Done()
+				for _, fetchedMetrics := range resp.Metrics {
+					for _, stats := range metric.Statistics {
+						id := fmt.Sprintf("id_%d", rand.Int())
+						period := int64(metric.Period)
+						mux.Lock()
+						getMetricDatas = append(getMetricDatas, cloudwatchData{
+							ID:                     resource.ID,
+							MetricID:               &id,
+							Metric:                 &metric.Name,
+							Service:                resource.Service,
+							Statistics:             []string{stats},
+							NilToZero:              &metric.NilToZero,
+							AddCloudwatchTimestamp: &metric.AddCloudwatchTimestamp,
+							Tags:                   metricTags,
+							Dimensions:             fetchedMetrics.Dimensions,
+							Region:                 &job.Region,
+							Period:                 &period,
+						})
+						mux.Unlock()
+					}
+				}
+			}
+		}()
+	}
+	wg.Wait()
+	maxMetricCount := 100
+	metricDataLength := len(getMetricDatas)
+	partition := int(math.Ceil(float64(metricDataLength) / float64(maxMetricCount)))
+	wg.Add(partition)
+	for i := 0; i < metricDataLength; i += maxMetricCount {
+		go func(i int) {
+			defer wg.Done()
```

Comment:
Also here the semaphore for Cloudwatch concurrency should be added:

``` 
		go func(i int) {
			defer wg.Done()
			cloudwatchSemaphore <- struct{}{}
			defer func() {
				<-cloudwatchSemaphore
			}()
			end := i + maxMetricCount
```

Reply:
why we need cloudwatchSemaphore here?
cloud watch get metrics data supports at least 25 calls per second and if I am not wrong cloudwatchSemaphore will restrict it to one API call at a time. which will increase Prometheus scrape time

Reply:
added Semaphore for tags